xmake/rules/c++/modules/xmake.lua

```diff
@@ -21,45 +21,122 @@
 -- define rule: c++.build.modules
 rule("c++.build.modules")
     set_extensions(".mpp", ".mxx", ".cppm", ".ixx")
+    add_deps("c++.build.modules.dependencies")
+    add_deps("c++.build.modules.builder")
+    add_deps("c++.build.modules.install")
+
     on_config(function (target)
         -- we disable to build across targets in parallel, because the source files may depend on other target modules
         -- @see https://github.com/xmake-io/xmake/issues/1858
-        local target_with_modules
-        for _, dep in ipairs(target:orderdeps()) do
-            local sourcebatches = dep:sourcebatches()
-            if sourcebatches and sourcebatches["c++.build.modules"] then
-                target_with_modules = true
-                break
-            end
-        end
+        local common = import("modules_support.common")
+
+        local target_with_modules = target:sourcebatches()["c++.build.modules"] and common.contains_modules(target) or false
         if target_with_modules then
             -- @note this will cause cross-parallel builds to be disabled for all sub-dependent targets,
             -- even if some sub-targets do not contain C++ modules.
             --
             -- maybe we will have a more fine-grained configuration strategy to disable it in the future.
             target:set("policy", "build.across_targets_in_parallel", false)
-            if target:has_tool("cxx", "clang", "clangxx") then
-                import("build_modules.clang").load_parent(target, opt)
-            elseif target:has_tool("cxx", "gcc", "gxx") then
-                import("build_modules.gcc").load_parent(target, opt)
-            elseif target:has_tool("cxx", "cl") then
-                import("build_modules.msvc").load_parent(target, opt)
-            else
-                local _, toolname = target:tool("cxx")
-                raise("compiler(%s): does not support c++ module!", toolname)
-            end
+
+            local modules_support = common.modules_support(target)
+
+            -- check C++20 module support
+            modules_support.check_module_support(target)
+
+            -- load parent
+            modules_support.load_parent(target, opt)
+
+            target:set("cxx.has_modules", true)
```

Comment:
we should use `target:data_set` and `target:data` instead of `target:set/get`, because it is only internal data.